README.md
# Breakthrough hospitalizations and deaths from COVID-19

Tools to extract and store official information on breakthrough hospitalizations anddeaths due to COVID-19 among the fully vaccinated population.

## IMPORTANT DISCLAIMER

***While these tools are for working with data published by official sources (currently only CDC), none of the work is endorsed by those organizations. I use organization names in modules/libraries so as an indicator of the source of data. Any opinion, commentary, inference etc expressed in the code and/or documentation is based on my personal perspective, does not imply endorsement by any organization or any current, past, or future employers/clients etc.***

## Extract and store CDC data on breakthrough hospitalizations and deaths from Covid-19

Currently, this is the only implementation.

The CDC provides some information on the number of hospitalizations and deaths from Covid-19 among those fully vaccinated against Covid-19 in the U.S. They do this by updating the contents of the [COVID-19 Vaccine Breakthrough Case Investigation and Reporting](https://www.cdc.gov/vaccines/covid-19/health-departments/breakthrough-cases.html) page instead of providing a proper time series in a proper data exchange file format.

I am interested specifically in this time series. Irregular snapshots of this page are currently [archived on archive.today](https://archive.is/https://www.cdc.gov/vaccines/covid-19/health-departments/breakthrough-cases.html). I downloaded those snapshots and manually created a couple of spreadsheets by copying and pasting numbers. That is error-prone so I decided to automate it.

Due to both changes introduced by the CDC to the layout of the page and the vagaries of dealing with archived snapshots, the code currently has to deal with four different types of pages with subtle differences in locating the information of interest. Thanks to the excellent [HTML::TreeBuilder::XPath](https://metacpan.org/pod/HTML::TreeBuilder::XPath) module, the code to extract the information was mostly straightforward.

Originally, this information was to be published every Friday for the period up to the previous Monday [as stated by the CDC](https://archive.ph/hAK1h):

> As CDC and state health departments shift to focus only on investigating vaccine breakthrough cases that result in hospitalization or death, those data will be regularly updated and posted every Friday.

The CDC skipped the update on the Friday immediately prior to the [approval of the Pfizer vaccine (Comirnaty)](https://www.fda.gov/news-events/press-announcements/fda-approves-first-covid-19-vaccine). The update on August 20th was delayed to August 23 and was released after the vaccine approval was announced. After a couple of updates being released on Mondays to reflect information up to the preceding Monday, on Friday, September 10, 2021, the CDC released data covering up to the preceding Tuesday. I am not sure what the schedule will be going forward.

As of September 10, 2021, [CDC continue to state](https://www.cdc.gov/vaccines/covid-19/health-departments/breakthrough-cases.html):

> Ultimately, CDC will use the [National Notifiable Diseases Surveillance System (NNDSS)](https://archive.ph/o/hAK1h/https://wwwn.cdc.gov/nndss/) to identify vaccine breakthrough cases.

I am not sure why this has not been implemented over the five months since mid-April this year. I hope that when it is implemented, the CDC provide more detailed information regarding breakthrough cases, hospitalizations, and deaths as a time series.

Note the CDC caution on selection bias in the data:

> It is important to note that reported vaccine breakthrough cases will represent an undercount. This surveillance system is passive and relies on voluntary reporting from state health departments which may not be complete.  Also, not all real-world breakthrough cases will be identified because of lack of testing. This is particularly true in instances of asymptomatic or mild illness.

It is important to note that most asymptotic or mild cases of COVID19 infection are unlikely to make it into official statistics either.

## INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

## SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc CDC::Covid19::Data::Breakthrough

## LICENSE AND COPYRIGHT

This software is Copyright © 2021 by A. Sinan Unur.

This is free software, licensed under:

  [The Apache License version 2.0](LICENSE)


Changes.md
# CHANGES

2021-09-10: First version that parses all known snapshots of CDC data


Makefile.PL
use v5.28;
use strict;
use warnings;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
    NAME             => 'CDC::Covid19::Data::Breakthrough',
    AUTHOR           => q{A. Sinan Unur <https://www.unur.com/contact.html>},
    VERSION_FROM     => 'lib/CDC/Covid19/Data/Breakthrough.pm',
    ABSTRACT_FROM    => 'lib/CDC/Covid19/Data/Breakthrough.pm',
    LICENSE          => 'artistic_2',
    MIN_PERL_VERSION => 'v5.28',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    TEST_REQUIRES => {
        'Test::More' => '0',
    },
    PREREQ_PM => {
        'Const::Fast' => '0',
        'HTML::TreeBuilder::XPath' => '0',
        'JSON::MaybeXS' => '0',
        'Path::Tiny' => '0',
        'Text::CSV_XS' => '0',
        'Text::Table::HTML' => '0',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'CDC-Covid19-Data-Breakthrough-*' },
);

# Compatibility with old versions of ExtUtils::MakeMaker
unless (eval { ExtUtils::MakeMaker->VERSION('6.64'); 1 }) {
    my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES} || {};
    @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

unless (eval { ExtUtils::MakeMaker->VERSION('6.55_03'); 1 }) {
    my $build_requires = delete $WriteMakefileArgs{BUILD_REQUIRES} || {};
    @{$WriteMakefileArgs{PREREQ_PM}}{keys %$build_requires} = values %$build_requires;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
    unless eval { ExtUtils::MakeMaker->VERSION('6.52'); 1 };
delete $WriteMakefileArgs{MIN_PERL_VERSION}
    unless eval { ExtUtils::MakeMaker->VERSION('6.48'); 1 };
delete $WriteMakefileArgs{LICENSE}
    unless eval { ExtUtils::MakeMaker->VERSION('6.31'); 1 };

WriteMakefile(%WriteMakefileArgs);


lib/CDC/Covid19/Data/Breakthrough.pm
package CDC::Covid19::Data::Breakthrough 0.001 {
    use utf8;
    use v5.28;
    use strict;
    use warnings;

    use feature 'signatures';
    no warnings 'experimental::signatures';

    use Carp qw( croak );
    use Const::Fast qw( const );
    use JSON::MaybeXS qw();

    # Until I figure out something that matches C but not D, put D before C.
    use constant LAYOUTS => qw(
        CDC::Covid19::Data::Breakthrough::LayoutA
        CDC::Covid19::Data::Breakthrough::LayoutB
        CDC::Covid19::Data::Breakthrough::LayoutE
        CDC::Covid19::Data::Breakthrough::LayoutD
        CDC::Covid19::Data::Breakthrough::LayoutC
    );

    BEGIN {
        for my $layout (LAYOUTS) {
            my $ret = eval "require $layout; 1";
            croak $@ unless $ret;
        }
    }

    # Use $VAR{xyz} instead of 'xyz'. Because `%VAR` is constant, typos manifest
    # as loud errors rather than silent problems that pop out much later.

    const my %VAR => map +($_ => $_), qw(
        deaths
        deaths_from_covid19
        hospitalized
        hospitalized_from_covid19
        date
        date_updated
        non_covid19_deaths
        non_covid19_hospitalized
        public_sources
        vaccinated
    );

    # Read only accessors for the properties in %VAR

    for my $v (keys %VAR) {
        no strict 'refs';
        *{$v} = sub { $_[0]->{data}{$v} };
    }

    sub new ($class, $tree) {
        my $self = {};

        $self->{layout} = $class->_detect($tree);

        bless $self => $class;

        $self->_extract($tree);

        return $self;
    }

    sub columns ($class) {
        [ grep !/public_sources/, keys %VAR ];
    }

    sub layout ($self) {
        $self->{layout};
    }

    sub to_json($self, %options) {
        my $json = JSON::MaybeXS->new(%options);
        $json->encode($self->{data});
    }

    sub _detect ($class, $tree) {
        my ($detected) = grep scalar $tree->findnodes($_->discriminator), $class->LAYOUTS;
        return $detected;
    }

    sub _extract ($self, $tree) {
        $self->_extract_date($tree);
        $self->_extract_date_updated($tree);
        $self->_extract_deaths($tree);
        $self->_extract_hospitalized($tree);
        $self->_extract_public_sources($tree);
        $self->_extract_vaccinated($tree);

        return;
    }

    sub _extract_date($self, $tree) {
        my $var = $VAR{date};
        $self->{data}{$var} = _yyyymmdd_of($tree->findvalue($self->layout->date));

        return;
    }

    sub _extract_date_updated ($self, $tree) {
        my $var = $VAR{date_updated};
        my $val = $tree->findnodes($self->layout->$var)->[0]->attr('content');

        $self->{data}{$var} = _yyyymmdd_of($val);
        return;
    }

    sub _extract_severe_cases ($self, $tree, $case_type) {
        my $var = $VAR{$case_type};
        my $non_covid19 = $VAR{"non_covid19_${case_type}"};

        for my $v ($var, $non_covid19) {
            $self->{data}{$v} = $tree->findvalue($self->layout->$v);
            $self->{data}{$v} =~ s/[^0-9]+//g;
            $self->{data}{$v} += 0;
        }

        $self->{data}{$VAR{"${case_type}_from_covid19"}} = $self->$case_type - $self->$non_covid19;

        return;
    }

    sub _extract_deaths ($self, $tree) {
        $self->_extract_severe_cases($tree, 'deaths');
    }

    sub _extract_hospitalized ($self, $tree) {
        $self->_extract_severe_cases($tree, 'hospitalized');
    }

    sub _extract_public_sources ($self, $tree) {
        my $var = $VAR{public_sources};

        $self->{data}{$var} = [
            map [
                map(s/external icon//r, $_->as_text),
                $_->attr('href')
            ], $tree->findnodes($self->layout->$var)
        ];

        return;
    }

    sub _extract_vaccinated ($self, $tree) {
        my $var = $VAR{vaccinated};

        $self->{data}{$var} = $tree->findvalue($self->layout->$var);
        $self->{data}{$var} =~ s/[^0-9]+//g;
        $self->{data}{$var} += 0;

        return;
    }

    sub _yyyymmdd_of ($val) {
        state %months = map {
            my $m = $_;
            map +($m->[$_] => 1 + $_), 0 .. $#$m;
        } [
            qw(
                January
                February
                March
                April
                May
                June
                July
                August
                September
                October
                November
                December
            )
        ];

        state $pat = qr{
            (?<month> @{[join '|', sort keys %months]}) \s+
            (?<day> [123]?[0-9]), \s+
            (?<year> 2021)
        }x;

        if ($_[0] =~ /^([^T]+)T/) {
            return $1;
        }

        if ($_[0] =~ $pat) {
            return sprintf '%04d-%02d-%02d', $+{year}, $months{$+{month}}, $+{day};
        }

        croak "Failed to parse date string '$_[0]'";
    }
=pod

=head1 NAME

CDC::Covid19::Data::Breakthrough - Extract and store Covid19 breakthrough hospitalizations/deaths data provided by the U.S. CDC

=head1 VERSION

Version 0.001

=head1 SYNOPSIS

This module provides a uniform interface to extract and store Covid-19 breakthrough hospitalization and death data provided by the U.S. CDC. The CDC updates the contents of the page L<COVID-19 Vaccine Breakthrough Case Investigation and Reporting|https://www.cdc.gov/vaccines/covid-19/health-departments/breakthrough-cases.html>. Since updates overwrite previous information, producing the time series relies on L<snapshots stored on archive.today|https://archive.is/https://www.cdc.gov/vaccines/covid-19/health-departments/breakthrough-cases.html>.

The constructor takes a single argument which specifies the URL or path to a snapshot of that URL. The module then invokes the correct extractor to read the data from the page and store it in a simple object that can be serialized as desired.

    use CDC::Covid19::Data::Breakthrough;

    my $data = CDC::Covid19::Data::Breakthrough->new($snapshot_file);
    ...

=head1 METHODS

=head2 C<new ($snapshot_file)>

This is the constructor. Invoke it with the path to a snapshot file.

=head2 C<columns>

List of data columns.

=head2 C<layout>

Detected layout for the snapshot file.

=head2 C<to_json>

Serialize to JSON.

=head2 C<date>

The end date of the period to which the data refer.

=head2 C<date_updated>

Contains archival date for snapshots obtained from L<archive.today|https://archive.today/https://www.cdc.gov/vaccines/covid-19/health-departments/breakthrough-cases.html>. Otherwise, the C<YYYY-MM-DD> formatted contents of the C<cdc:last_updated> meta tag.

=head2 C<deaths>

Number of fully vaccinated people who died with a positive Covid-19 test result.

=head2 C<deaths_from_covid19>

C<deaths> - C<non_covid19_deaths>.

=head2 C<hospitalized>

Number of fully vaccinated people who are hospitalized with a positive Covid-19 test result.

=head2 C<hospitalized_from_covid19>

C<hospitalized> - C<non_covid19_hospitalized>.

=head2 C<non_covid19_deaths>

Number of people who had no symptoms of COVID-19 or whose death was not COVID-related.

=head2 C<non_covid19_hospitalized>

Number of people who had no symptoms of COVID-19 or whose hospitalization was not COVID-related.

=head2 C<public_sources>

States and jurisdictions that are publicly reporting vaccine breakthrough cases identified by CDC.

=head2 C<vaccinated>

Number of people who received their second vaccine dose at least two weeks prior in millions.

=head1 AUTHOR

L<A. Sinan Unur|https://www.unur.com/contact.html>

=head1 BUGS

Please report any bugs or feature requests using L<GitHub issues|https://github.com/nanis/cdc-covid19-breakthrough-data/issues>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc CDC::Covid19::Data::Breakthrough

=head1 ACKNOWLEDGEMENTS

I ❤️ Perl and owe thanks to both Perl5 porters and the developers who maintain libraries on L<CPAN|https://metacpan.org>. As always, all errors and bugs are mine.

=head1 LICENSE AND COPYRIGHT

This software is Copyright (c) 2021 by A. Sinan Unur.

This is free software, licensed under:

  The Apache License version 2.0

=cut
    __PACKAGE__;
}
__END__


lib/CDC/Covid19/Data/Breakthrough/LayoutA.pm
package CDC::Covid19::Data::Breakthrough::LayoutA 0.001 {
    use utf8;
    use v5.28;
    use strict;
    use warnings;

    use feature 'signatures';
    no warnings 'experimental::signatures';

    use constant date => '//li[starts-with(normalize-space(text()), "As of")]';

    use constant date_updated => '/html/head/meta[@property="article:modified_time"]';

    use constant deaths => sprintf(
        'substring-before(substring-after(normalize-space(string(//li[contains(text(), "%s")])), "%s"), "%s")',
        'people with breakthrough infections were known to be hospitalized',
        'and',
        ' (',
    );

    use constant hospitalized => sprintf(
        'substring-before(normalize-space(string(//li[contains(text(), "%s")])), "%s")',
        'people with breakthrough infections were known to be hospitalized',
        ' ',
    );

    use constant non_covid19_deaths => sprintf(
        # Example: We want to extract the '9' in the following sentence:
        # > Of the 74 fatal cases, 9 (12%) were reported as asymptomatic or the patient died due to a cause not related to COVID-19.
        'substring-before(substring-after(normalize-space(string(//li[contains(text(), "%s")])), "%s"), "%s")',
        'died due to a cause not related to COVID-19',
        'fatal cases, ',
        ' (',
    );

    use constant non_covid19_hospitalized => sprintf(
        # Example: We want to extract the '133' in the following sentence:
        # > Of the 396 hospitalized patients, 133 (34%) were reported as asymptomatic or hospitalized for a reason not related to COVID-19.
        'substring-before(substring-after(normalize-space(string(//li[contains(text(), "%s")])), "%s"), "%s")',
        'hospitalized for a reason not related to COVID-19',
        'hospitalized patients, ',
        ' (',
    );

    use constant vaccinated => sprintf(
        'substring-before(substring-after(normalize-space(string(//a[contains(text(), "%s")])), "%s"), "%s")',
        'million people',
        'more than',
        'million people',
    );

    use constant public_sources => '/doesnotexist';

    use constant discriminator => sprintf(
        '//h2[normalize-space(text())="%s"]',
        'COVID-19 vaccine breakthrough infections reported to CDC',
    );

    __PACKAGE__;
}
__END__



lib/CDC/Covid19/Data/Breakthrough/LayoutC.pm
package CDC::Covid19::Data::Breakthrough::LayoutC 0.001 {
    use utf8;
    use v5.28;
    use strict;
    use warnings;

    use feature 'signatures';
    no warnings 'experimental::signatures';

    use parent 'CDC::Covid19::Data::Breakthrough::LayoutB';

    use constant date => sprintf(
        'substring-after(normalize-space(string(//h2[contains(normalize-space(text()), "%s")])), "%s")',
        'reported to CDC as of',
        'as of ',
    );

    # Example for the following expressions:
    # > *342 (30%) of 1,136 hospitalizations reported as asymptomatic or not related to COVID-19.
    # > †42 (18%) of 223 fatal cases reported as asymptomatic or not related to COVID-19.
    use constant deaths => sprintf(
        # We want to extract the '223' between 'of' and 'fatal cases' in the example above
        q{
            substring-before(
                substring-after(
                    substring-after(
                        normalize-space(string(//div[contains(normalize-space(text()), "%s")])),
                    "%s"),
                "%s"),
            "%s")
        },
        'reported as asymptomatic or not related to COVID-19',
        '†',
        'of ',
        ' fatal cases',
    );

    use constant hospitalized => sprintf(
        'substring-after(substring-before(normalize-space(string(//div[starts-with(text(), "%s")])), "%s"), "%s")',
        '*',
        'hospitalizations reported',
        'of',
    );

    use constant non_covid19_deaths => sprintf(
        # We want to extract the '42' between the dagger and the opening parenthesis in the example above
        q{
            substring-before(
                substring-after(
                    normalize-space(string(//div[contains(normalize-space(text()), "%s")])),
                "%s"),
            "%s")
        },
        'reported as asymptomatic or not related to COVID-19',
        '†',
        ' (',
    );

    use constant non_covid19_hospitalized => sprintf(
        'substring-after(substring-before(normalize-space(string(//div[starts-with(text(), "%s")])), "%s"), "%s")',
        '*',
        '(',
        '*',
    );

    use constant discriminator => sprintf(
        '//h2[starts-with(normalize-space(text()), "%s")]',
        'Hospitalized or fatal COVID-19 vaccine breakthrough cases reported to CDC as of',
    );

    __PACKAGE__;
}
__END__



lib/CDC/Covid19/Data/Breakthrough/LayoutD.pm
package CDC::Covid19::Data::Breakthrough::LayoutD 0.001 {
    use utf8;
    use v5.28;
    use strict;
    use warnings;

    use feature 'signatures';
    no warnings 'experimental::signatures';

    use constant date => sprintf(
        'substring-after(normalize-space(string(//h2[contains(normalize-space(text()), "%s")])), "%s")',
        'Hospitalized or fatal COVID-19 vaccine breakthrough cases reported to CDC as of ',
        'as of',
    );

    use constant date_updated => '/html/head/meta[@property="cdc:last_updated"]';

    use constant deaths => '//th[@id="deaths"]';

    use constant hospitalized => '//th[@id="hospitalized"]';

    use constant non_covid19_deaths => '//td[@headers="asymptomatic deaths"][1]';

    use constant non_covid19_hospitalized => '//td[@headers="asymptomatic hospitalized"][1]';

    use constant vaccinated => sprintf(
        q{
            substring-before(
                substring-after(
                    normalize-space(string(//a[starts-with(normalize-space(text()), "%s")])),
                "%s"),
            "%s")
        },
        'more than ',
        'more than ',
        ' million people',
    );

    use constant public_sources => '//div[@id="acc-panel-1"]//a[@class="tp-link-policy"]';

    use constant discriminator => '//table[@class="table table-bordered"]';

    __PACKAGE__;
}
__END__


lib/CDC/Covid19/Data/Breakthrough/LayoutB.pm
package CDC::Covid19::Data::Breakthrough::LayoutB 0.001 {
    use utf8;
    use v5.28;
    use strict;
    use warnings;

    use feature 'signatures';
    no warnings 'experimental::signatures';

    use constant date => '//div[starts-with(normalize-space(text()), "As of")]';

    use constant date_updated => '/html/head/meta[@property="article:modified_time"]';

    # Example for the following expressions:
    # > *167 (34%) of the 498 hospitalizations were reported as asymptomatic or not related to COVID-19.
    # > †11 (13%) of the 88 fatal cases were reported as asymptomatic or not related to COVID-19

    use constant deaths => sprintf(
        # We want to extract the '88' between 'of the' and 'fatal cases' in the example above
        q{
            substring-before(
                substring-after(
                    substring-after(
                        normalize-space(string(//div[contains(normalize-space(text()), "%s")])),
                    "%s"),
                "%s"),
            "%s")
        },
        'were reported as asymptomatic or not related to COVID-19',
        '†',
        'of the ',
        ' fatal cases',
    );

    use constant hospitalized => sprintf(
        'substring-after(substring-before(normalize-space(string(//div[starts-with(text(), "%s")])), "%s"), "%s")',
        '*',
        'hospitalizations were reported',
        'of the',
    );

    use constant non_covid19_deaths => sprintf(
        # We want to extract the '11' between the dagger and the opening parenthesis in the example above
        q{
            substring-before(
                substring-after(
                    normalize-space(string(//div[contains(normalize-space(text()), "%s")])),
                "%s"),
            "%s")
        },
        'were reported as asymptomatic or not related to COVID-19',
        '†',
        ' (',
    );


    use constant non_covid19_hospitalized => sprintf(
        'substring-after(substring-before(normalize-space(string(//div[starts-with(text(), "%s")])), "%s"), "%s")',
        '*',
        '(',
        '*',
    );

    use constant vaccinated => 'string(//a[contains(normalize-space(text()), "More than")])';

    use constant public_sources => '/doesnotexist';

    use constant discriminator => sprintf(
        '//h2[starts-with(normalize-space(text()), "%s")]',
        'COVID-19 vaccine breakthrough infections reported to CDC as of',
    );

    __PACKAGE__;
}
__END__
